packages/service-library/src/servicelib/aiohttp/long_running_tasks/client.py

```diff
@@ -0,0 +1,155 @@
+                if not task_status.done:
+                    raise TryAgain(
+                        f"{task_id=}, {task_status.started=} has "
+                        f"status: '{task_status.task_progress.message}'"
+                        f" {task_status.task_progress.percent}%"
+                    )
+    except TryAgain as exc:
+        # this is a timeout
+        raise asyncio.TimeoutError(
+            f"Long running task {task_id}, calling to {status_url} timed-out after {wait_timeout_s} seconds"
+        ) from exc
+
+
+@retry(**_DEFAULT_AIOHTTP_RETRY_POLICY)
+async def _task_result(session: ClientSession, result_href: str) -> Any:
+    async with session.get(result_href) as response:
+        response.raise_for_status()
+        if response.status != web.HTTPNoContent.status_code:
+            data, error = unwrap_envelope(await response.json())
+            assert not error  # nosec
+            assert data  # nosec
+            return data
+
+
+@retry(**_DEFAULT_AIOHTTP_RETRY_POLICY)
+async def _abort_task(session: ClientSession, abort_href: str) -> None:
+    async with session.delete(abort_href) as response:
+        response.raise_for_status()
+        data, error = unwrap_envelope(await response.json())
+        assert not error  # nosec
+        assert not data  # nosec
+
+
+@dataclass(frozen=True)
+class LRTask:
+    progress: TaskProgress
+    _result: Optional[Coroutine[Any, Any, Any]] = None
+
+    def done(self) -> bool:
+        return self._result is not None
+
+    async def result(self) -> Any:
+        if not self._result:
+            raise ValueError("No result ready!")
+        return await self._result
+
+
+async def long_running_task_request(
+    session: ClientSession,
+    url: URL,
+    json: Optional[RequestBody] = None,
+    wait_timeout_s: int = 1 * _HOUR,
+    wait_interval_s: float = 1,
+) -> AsyncGenerator[LRTask, None]:
+    """Will use the passed `ClientSession` to call an oSparc long
+    running task `url` passing `json` as request body.
+    NOTE: this follows the usual aiohttp client syntax, and will raise the same errors
+
+    :param session: The client to use
+    :type session: ClientSession
+    :param url: The url to call. NOTE: the endpoint must follow oSparc long running task server syntax (202, then status, result urls)
+    :type url: URL
+    :param json: optional body as dictionary, defaults to None
+    :type json: Optional[RequestBody], optional
+    :param wait_timeout_s: after this timeout is reached and no result is ready will raise asyncio.TimeoutError, defaults to 1*_HOUR
+    :type wait_timeout_s: int, optional
+    :param wait_interval_s: will check for result every `wait_interval` seconds, defaults to 1
+    :type wait_interval_s: float, optional
+    :return: a task containing the final result
+    :rtype: AsyncGenerator[LRTask, None]
+    :yield: a task containing the current TaskProgress
+    :rtype: Iterator[AsyncGenerator[LRTask, None]]
+    """
+    task = None
```

Comment:
So, this function encapsulates logic with four possible requests, namely
- ``_start``
- ``_waif_for_completion``
- ``_task_result``
- ``_abort_task``

Each of the above can potentially raise [``ClientResponseError``](https://docs.aiohttp.org/en/stable/client_reference.html#aiohttp.ClientResponseError), therefore the exceptions interface here remains ``ClientResonseError``. **Q1**: how does the caller know which of these requests above errored? Would it make sense to define an error wrapper that associates ``ClientResonseError`` upon ``start``, ``while completion``, etc??

Reply:
hmm, I think the only where it might make sense to differentiate is the start. the rest shall behave as usual. as this is wrapped, I do not think the caller might know what is going on with the long running task internals.